Add mount options e2e test

wongma7 opened the pull request:
**What this PR does / why we need it**: A test for newly added StorageClass.mountOptions and PV.mountOptions: provision a pv using a class with its storageclass.mountoptions set, and the end result should be that the mount options can be seen from the mounter.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: Fixes #52138

**Special notes for your reviewer**: 

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

Commit:
Add mount options e2e test

wongma7 commented:
/sig storage

jeffvance commented:
/test pull-kubernetes-e2e-gce-bazel

jeffvance commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -126,8 +127,15 @@ func testDynamicProvisioning(t storageClassTest, client clientset.Interface, cla
+	// been mounted with the PV's mount options, if the PV was provisioned with any
+	for _, option := range pv.Spec.MountOptions {
+		// -M to define mountpoint, -o to define output column, -n to exclude header
+		command += fmt.Sprintf(" && [[ \",`findmnt -M /mnt/test/data -o OPTIONS -n`,\" = *\",%s,\"* ]]", option)
```
Is there a mount options delimiter you can include in the `findmnt` test? Eg. if a mount option is a subset of another mount option you'll get a false positive.  Eg: `atime` is a subset of `noatime`

jeffvance replied:
It probably doesn't matter here but could be more efficient to capture the `fndmnt` options once and then do a bash "includes" string compare for each class option

wongma7 replied:
each option is surrounded by commas (we add commas to the beginning and end of findmnt), so it will work because it's checking if the full ",noatime," exists.

yeah I just wanted the simplest/shortest bash expression so settled on this
 https://stackoverflow.com/questions/29300806/a-bash-script-to-check-if-a-string-is-present-in-a-comma-separated-list-of-strin

jeffvance replied:
the `-M` option does not appear to be supported in rhel, centos, fedora images. It is supported in ubuntu. It looks like the `--target` findmnt option is more universally supported.

jeffvance commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -408,6 +416,29 @@ var _ = SIGDescribe("Dynamic Provisioning", func() {
+				},
+			}
+			class := newStorageClass(test, ns, "mountoptions")
+			class.MountOptions = []string{"debug"}
```
is it worth specifying several mount options here?

wongma7 replied:
sure, added resgid=0, should be harmless.

jeffvance commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -126,8 +127,15 @@ func testDynamicProvisioning(t storageClassTest, client clientset.Interface, cla
+	// been mounted with the PV's mount options, if the PV was provisioned with any
+	for _, option := range pv.Spec.MountOptions {
+		// -M to define mountpoint, -o to define output column, -n to exclude header
+		command += fmt.Sprintf(" && [[ \",`findmnt -M /mnt/test/data -o OPTIONS -n`,\" = *\",%s,\"* ]]", option)
```
ya, as I was editing my initial comment I finally noticed that you did surround both the individual mount option and the list of mount options with a comma... so I deleted that comment

wongma7 replied:
busybox doesn't have findmnt at all in fact, I have to either change the image to something that does or parse output of `mount`. I will try the latter, so much for simple and short bash :(.

wongma7 replied:
new expression with accompanying comment:
[[ "`mount | grep 'on /mnt/test/data' | awk '{print $6}' | sed 's/^(/,/; s/)$/,/'`" = *",%s,"* ]]

wongma7 commented:
My internet is lacking atm so I need to wait for CI to verify the test unfortunately. ofc I test what i can locally, ilke the mount parsing bash

jeffvance commented:
/test pull-kubernetes-e2e-gce-bazel
/test pull-kubernetes-e2e-gce-etcd3
/test pull-kubernetes-kubemark-e2e-gce

saad-ali commented:
/test pull-kubernetes-e2e-gce-bazel
/test pull-kubernetes-e2e-gce-etcd3

jeffvance commented:
/test pull-kubernetes-e2e-gce-bazel
/test pull-kubernetes-e2e-gce-etcd3
(seeing out of disk space errors)

saad-ali commented:
@wongma7 looks like these tests have issues, PTAL

wongma7 commented:
I've verified the tests are working now, PTAL @jeffvance 

two issues I fixed:
1. the arbitrarily chosen second mount option resgid=n wasn't applying, so i switched to nouid32 for the test
2. my bash was fine but busybox is running sh which didn't like the fancy if statement and globbing, so the new expression is just a grep.

jeffvance commented:
/lgtm

saad-ali commented:
/lgtm
/approve
/test pull-kubernetes-unit
/test pull-kubernetes-kubemark-e2e-gce

saad-ali commented:
/test pull-kubernetes-kubemark-e2e-gce

apelisse commented:
Needed for 1.8. Adding milestone.

apelisse commented:
I've updated the body of the PR so that Github closes the associated issue when merged

saad-ali commented:
Looks like this needs a rebase. @wongma7 can you please rebase?

wongma7 commented:
rebased

dims commented:
@saad-ali can you please re-add the lgtm?

apelisse commented:
Nothing looks terribly wrong to me ...
/lgtm
